core/src/main/java/org/mule/processor/chain/NonBlockingProcessorIterator.java

```diff
@@ -0,0 +1,98 @@
+import org.mule.api.MuleEvent;
+import org.mule.api.MuleException;
+import org.mule.api.processor.InterceptingMessageProcessor;
+import org.mule.api.processor.MessageProcessor;
+import org.mule.api.processor.MessageRouter;
+import org.mule.api.transport.NonBlockingResponseReplyToHandler;
+import org.mule.api.transport.ReplyToHandler;
+import org.mule.execution.MessageProcessorExecutionTemplate;
+import org.mule.api.processor.CompletionHandler;
+import org.mule.processor.NonBlockingMessageProcessor;
+
+import java.util.List;
+
+/**
+ * Specialized {@link org.mule.processor.chain.ProcessorIterator} that pauses iteration in the case a {@link org.mule.processor.NonBlockingMessageProcessor}
+ * is invoked and the flow is executing using non-blocking.  Processor iteration is then continued when the {@link org.mule.processor.NonBlockingMessageProcessor}
+ * invokes the {@link org.mule.api.transport.ReplyToHandler}.
+ */
+public class NonBlockingProcessorIterator extends ProcessorIterator
+{
+
+    private ReplyToHandler replyToHandler;
+
+    public NonBlockingProcessorIterator(MuleEvent event, List<MessageProcessor> processors,
+                                        MessageProcessorExecutionTemplate executionTemplate, boolean copyOnVoidEvent)
+    {
+        super(event, processors, executionTemplate, copyOnVoidEvent);
+        this.replyToHandler = event.getReplyToHandler();
+    }
+
+    @Override
+    protected void preProcess(MessageProcessor processor)
+    {
+        if (event.getExchangePattern().hasResponse() && !event.isSynchronous())
+        {
+            if ((processor instanceof MessageRouter || processor instanceof InterceptingMessageProcessor) && !
+                    (processor instanceof NonBlockingSupported))
+            {
+                event = new DefaultMuleEvent(event.getMessage(), event, true);
+            }
+
+            if (processor instanceof NonBlockingMessageProcessor)
+            {
+                event = new DefaultMuleEvent(event, new NonBlockingResponseReplyToHandler(new CompletionHandler<MuleEvent, MessagingException>()
+                {
+                    @Override
+                    public void onCompletion(MuleEvent o)
+                    {
+                        NonBlockingProcessorIterator.this.event = new DefaultMuleEvent(o, replyToHandler);
+                        try
+                        {
+                            resume();
+                        }
+                        catch (MessagingException e)
+                        {
+                            onFailure(e);
+                        }
+                    }
+
+                    @Override
+                    public void onFailure(MessagingException exception)
+                    {
+                        replyToHandler.processExceptionReplyTo(event, exception, null);
+                    }
+                }));
+            }
+        }
+    }
+
+    private void resume() throws MessagingException
```

Comment:
Let's discuss this one if you think it's importante.  I think it's harder to read putting it anonymous class and also the it's the iterator thats being resumed, not the anonymous class.  So it it was to go in callback then it should be renamed to resumeIterator or something.  I'm not convinced though.